### ISSUE TITLE
Loss not dropping on custom dataset :(

### ISSUE THREAD

**Issue by tuttelikz** [facebookresearch/dino#61]
Hi, thanks for the wonderful work, @mathildecaron31!
Reported video is inspiring :D
__
I am experimenting with a custom dataset.
The thing is, it's totally okay to train vision transformer (`deit_small`) in supervised manner and loss drops fine.
Even managed to apply `visualize_attention.py` to see heatmaps for a separately trained ViT.
But when I switch to use self-supervised Dino setup, there is almost no change in loss during training.
Do you have idea why it could happen or possible solutions?
__
Thanks!

I am attaching screenshot from training and arguments I have used for training script.

![loss-stop](https://user-images.githubusercontent.com/6753761/120320338-05943900-c31d-11eb-87fb-852faac1b772.png)

```python
arch ='deit_small'
patch_size = 16
out_dim = 10000 # default 65536
norm_last_layer = False
momentum_teacher = 0.996 # check this according to batch_size
bsize = 256 #####
use_bn_in_head = False
warmup_teacher_temp = 0.0005 # less if does not decrease, default 0.04
teacher_temp = 0.3 # increase if needed, default: 0.04
warmup_teacher_temp_epochs = 0 # default 30 to warmup
use_fp16 = False #disable is loss is unstable, default: True
weight_decay = 0.04 # a smaller value works well
weight_decay_end = 0.4 # final value of weight decay
clip_grad = 3.0 # max parameter gradient norm, 0 for disabling # default, 3.0
batch_size_per_gpu = 256 # reduce this if not fit, default 64
epochs = 100
freeze_last_layer = 5 # default 1, Try increasing this value if the loss does not decrease.
lr = 0.005 #linear with batch size scaled, for ref of 256, def 0.0005
warmup_epochs = 0 #linear warmup def 10
min_lr = 1e-6 # target lr at the optimization
optimizer = 'sgd' # def: adamw
global_crops_scale = (0.4, 1.)
local_crops_number = 8 # local small views
local_crops_scale = (0.05, 0.4) # def (0.05, 0.4)
data_path = train_dataset_dir #
output_dir = "./dirlog"
saveckp_freq = 20
seed = 0 # random seed
num_workers = 40 #def:10
dist_url = "env://"
local_rank = 0
device_ids = [0, 1, 2, 3, 4, 5] # use 6 gpus
```

**Comment by mathildecaron31** [comment]
Hi @tuttelikz 
Thanks for your kind words.
Can you try the following for improving the stability:

* `--norm_last_layer true` (This will l2 normalize the last layer weights)
* `--use_fp16 false` (But you're already doing that :) )
* `--optimizer adamw` (Is there a motivation for using sgd instead of adamw ?) and thus adapt the learning rate as before. Also if you choose to use sgd it is posible that you need to re-adapt the weight decay (maybe use a much lower value). I'd recommend starting from the default optim params with adamw.

If I understand correctly the effective batch size is 1536 (256 * 6). Can you try reducing that a bit ? I've observed that large batch training can be unstable.

Hope that helps.

**Comment by black0017** [comment]
@mathildecaron31  @tuttelikz

I am seeing a similar behavior (loss not decreasing) on smaller datasets (~12K images) both with resnet50 and "vit_small"/"deit_small"

I am actually looking to reproduce the results in a smaller scale that would fit an academic budget compute power like 2 gpus.

Btw, I would change the teacher's  `teacher_temp = 0.3` to 0.04 or 0.02

Intuitively, smaller temp -> sharper predictions.

I find it difficult to drop the initial loss which in my mind translates to the fact that the teacher is not  better at the student in the beginning of training.  I have also found the `out_dim` parameter which is actually the soft classes to be highly related to the datasets size.
 
If you have any updates on this please let us know.
 
Best,

N.

**Comment by mathildecaron31** [comment]
Hi,

Yes the teacher temp `0.3` feels to high, a value between 0.02 and 0.04 would be more suited.

I agree that you should reduce `out_dim` for smaller dataset.

**Comment by mathildecaron31** [comment]
Hi @tuttelikz , any update ? have you been able to make the model converge ?

**Comment by hayatrajani** [comment]
On similar lines, I too am experimenting with a custom dataset (of about 70k images) and would like to share my results. The loss ([logs.txt](https://github.com/facebookresearch/dino/files/7038661/logs.txt)) does decrease with all the default hyperparameter settings except the following changes:

- `use_norm_layer=True`
- `out_dim=8192`
- `lr=5e-2`
- `min_lr=5e-4`

I have one question though with respect to this and the training logs provided in the repo. In the training logs, the loss does decrease initially but begins to rise again in later epochs. But the model still shows better performance. Doesn't the "lower the loss, better the performance" convention apply in this case? How can one then tell if the model converges?

Thanks!!

**Comment by mathildecaron31** [comment]
HI @hayatrajani 

Yes, we have indeed observed that a lower training loss does not usually lead to better transfer performance.

That is why in unsupervised learning I think it is important to monitor the performance by transferring to a downstream task rather than just looking at the training loss.

In DINO trainings case I think this can be explained by the fact that a lot of things are moving along training (i.e. the targets with the momentum teacher, the regularization strength, etc).

**Comment by black0017** [comment]
@hayatrajani  i think you refer to the warmup stage of training where many stuff are happening:

- teacher temperature goes up for warmup_teacher_epochs
- learning rate & weight decay change up for warmup epochs 
- the last linear layer if freeze for some epochs also

Here is an indicative picture of one of my small scale experiments on cifar10

![image](https://user-images.githubusercontent.com/13345132/130613578-6ca75f26-c19a-44c5-83fb-8feab29b2d49.png)

In general I would advice to let the training for many epochs (even >300 for smaller datasets like cifar10 and stl10). Of course you have to cross check the K-NN accuracy with the different checkpoints as @mathildecaron31 says

**Comment by hayatrajani** [comment]
Thanks a lot @mathildecaron31 @black0017 !!

This sure does help clear up a couple things.

Best,
Hayat

**Comment by mathildecaron31** [comment]
Best way to thank me is to star the repo haha :D

**Comment by mathildecaron31** [comment]
Closing the issue. Feel free to re-open if needed.

cc @tuttelikz

**Comment by TonyUSTC** [comment]
@tuttelikz @hayatrajani  I get same problem, so how to slove this problem?
